lib/chef/provider/launchd.rb

```diff
@@ -0,0 +1,198 @@
+        @group = @new_resource.group
+        @mode = @new_resource.mode
+        @owner = @new_resource.owner
+        @path = @new_resource.path
+        @source = @new_resource.source
+        @content = content?
+      end
+
+      def action_create
+        manage_plist(:create)
+      end
+
+      def action_create_if_missing
+        manage_plist(:create_if_missing)
+      end
+
+      def action_delete
+        # If you delete a service you want to make sure its not loaded or
+        # the service will be in memory and you wont be able to stop it.
+        if ::File.exists?(@path)
+          manage_service(:disable)
+        end
+        manage_plist(:delete)
+      end
+
+      def action_enable
+        if manage_plist(:create)
+          manage_service(:restart)
+        else
+          manage_service(:enable)
+        end
+      end
+
+      def action_disable
+        manage_service(:disable)
+      end
+
+      def manage_plist(action)
+        if @source
+          res = cookbook_file_resource
+        else
+          res = file_resource
+        end
+        res.run_action(action)
+        new_resource.updated_by_last_action(true) if res.updated?
+        res.updated
+      end
+
+      def manage_service(action)
+        res = service_resource
+        res.run_action(action)
+        new_resource.updated_by_last_action(true) if res.updated?
+      end
+
+      def service_resource
+        res = Chef::Resource::MacosxService.new(@label, run_context)
+        res.name(@label)
+        res.service_name(@label)
+        res.plist(@path)
+        res.session_type(@session_type) if @session_type
+
+        res
+      end
+
+      def file_resource
+        res = Chef::Resource::File.new(@path, run_context)
+        res.name(@path)
+        res.backup(@backup) if @backup
+        res.content(@content) if @content
+        res.group(@group) if @group
+        res.mode(@mode) if @mode
+        res.owner(@owner) if @owner
+
+        res
+      end
+
+      def cookbook_file_resource
+        res = Chef::Resource::CookbookFile.new(@path, run_context)
+        res.cookbook_name = @cookbook if @cookbook
+        res.name(@path)
+        res.backup(@backup) if @backup
+        res.group(@group) if @group
+        res.mode(@mode) if @mode
+        res.owner(@owner) if @owner
+        res.source(@source) if @source
+
+        res
+      end
+
+      def define_resource_requirements
+        requirements.assert(
+          :create, :create_if_missing, :delete, :enable, :disable
+        ) do |a|
+          type = @new_resource.type.to_s
+          a.assertion { ['daemon', 'agent'].include?(type) }
+          error_msg = 'type must be daemon or agent'
+          a.failure_message Chef::Exceptions::ValidationFailed, error_msg
+        end
+      end
+
+      def content?
+        if @new_resource.hash
+          @new_resource.hash.to_plist
+        elsif @new_resource.program || @new_resource.program_arguments
+          gen_plist
+        end
```

Comment:
so:

gen_plist should probably just return a hash so that we can call to_plist in one spot here.  this also seems to be overloaded as a boolean check even though it returns a plist.  plus we should explicitly require plist in this file, and we should not use the monkeypatch (really whatever introduced that `Hash#to_plist` monkeypatch should get ripped out).

```ruby
require 'plist'

def content?
  !!content
end

def content
  plist_hash = new_resource.hash || gen_plist_hash
  Plist::Emit.dump(plist_hash) unless plist_hash.nil? 
end

def gen_plist_hash
  return nil unless @new_resource.program || @new_resource.program_arguments
  [...exactly like gen_plist but returns a hash...]
end
```